apollo-router/src/plugins/file_uploads/mod.rs

```diff
@@ -245,9 +269,188 @@ type MapPerVariable = HashMap<String, HashMap<String, Vec<Vec<String>>>>;
 #[derive(Clone)]
 struct SupergraphLayerResult {
     multipart: Arc<Mutex<Multipart<'static>>>,
+    files_order: IndexSet<String>,
     map_per_variable: MapPerVariable,
 }
 
+// FIXME: Remove async???
+async fn rearange_execution_plan(mut req: execution::Request) -> execution::Request {
+    let supergraph_result = req
+        .context
+        .private_entries
+        .lock()
+        .get::<SupergraphLayerResult>()
+        .cloned();
+    if let Some(supergraph_result) = supergraph_result {
+        let SupergraphLayerResult {
+            files_order,
+            map_per_variable,
+            ..
+        } = supergraph_result;
+
+        let root = &req.query_plan.root;
+        let (_, root) = rearrange_plan_node(root, &files_order, &map_per_variable);
+        req = execution::Request {
+            query_plan: Arc::new(QueryPlan {
+                root,
+                usage_reporting: req.query_plan.usage_reporting.clone(),
+                formatted_query_plan: req.query_plan.formatted_query_plan.clone(),
+                query: req.query_plan.query.clone(),
+            }),
+            ..req
+        };
+    }
+    req
+}
+
+fn rearrange_plan_node<'a>(
```

Comment:
This logic may need to be moved to allow the resulting query plan to be cached.

Reply:
We discussed and we're not going to do this for preview as there are other things that need to be taken into account such as the contents of the file map on the second part of the stream. We will tackle for GA though.